oauth2client/django_util/views.py

```diff
@@ -0,0 +1,109 @@
+# Copyright 2015 Google Inc.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import hashlib
+import json
+import os
+import pickle
+
+from django.core.urlresolvers import reverse
+from django.http import HttpResponseBadRequest
+from django.shortcuts import redirect
+from oauth2client.client import FlowExchangeError, OAuth2WebServerFlow
+from oauth2client.django_util import oauth2
+from oauth2client.django_util.signals import oauth2_authorized
+from oauth2client.django_util.storage import get_storage
+
+
+def _make_flow(request, scopes, return_url=None):
+    """Creates a Web Server Flow"""
+    # Generate a CSRF token to prevent malicious requests.
+    csrf_token = hashlib.sha256(os.urandom(1024)).hexdigest()
+
+    request.session['google_oauth2_csrf_token'] = csrf_token
+
+    state = json.dumps({
+        'csrf_token': csrf_token,
+        'return_url': return_url,
+    })
+
+    flow = OAuth2WebServerFlow(
+        client_id=oauth2.client_id,
+        client_secret=oauth2.client_secret,
+        scope=scopes,
+        state=state,
+        redirect_uri=request.build_absolute_uri(
+            reverse("google_oauth:callback")))
+
+    flow_key = 'google_oauth2_flow_{0}'.format(csrf_token)
+    request.session[flow_key] = pickle.dumps(flow)
+    return flow
+
+
+def _get_flow_for_token(csrf_token, request):
+    flow_pickle = request.session.get(
+        'google_oauth2_flow_{0}'.format(csrf_token), None)
+    if not flow_pickle:
+        return
+    return pickle.loads(flow_pickle)
+
+
+def oauth2_callback(request):
+    if 'error' in request.GET:
+        reason = request.GET.get('error_description',
+                                 request.GET.get('error', ''))
+        return HttpResponseBadRequest(
+            'Authorization failed failed: %s' % reason)
+
+    encoded_state = request.GET.get('state')
```

Comment:
Here, you ask for defaults then check for bad state, but later you just try to get the items then catch exceptions. I'd prefer if you make this a `try...except` for consistency.